catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,285 @@
+        currentSourceKeyAndSourceMap.entrySet()) {
+      final SourceKey sourceKey = sourceKeySourceEntry.getKey();
+      final Source source = sourceKeySourceEntry.getValue();
 
-    scheduler.shutdownNow();
+      final Future<T> future;
+      try {
+        future = pollThreadPool.submit(() -> getValue(source));
+      } catch (RejectedExecutionException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      }
+
+      final Callable<Runnable> callable =
+          () -> {
+            try {
+              final T newValue = future.get(timeout, timeoutTimeUnit);
+              if (newValue == null) {
+                throw new IllegalArgumentException("getValueFunction may not return null");
+              }
+
+              return () -> cacheNewValue(sourceKey, newValue);
+            } catch (TimeoutException e) {
+              return () -> handleTimeout(sourceKey);
+            } catch (ExecutionException e) {
+              final Throwable cause = e.getCause();
+              if (cause instanceof RuntimeException) {
+                return () -> handleException(sourceKey, (RuntimeException) cause);
+              } else {
+                // TODO
+                throw new IllegalStateException("TODO");
+              }
+            } catch (InterruptedException e) {
+              // TODO
+              throw new IllegalStateException("TODO");
+            }
+          };
+
+      try {
+        completionService.submit(callable);
+      } catch (RejectedExecutionException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      }
+    }
+
+    for (int i = 0; i < currentSourcesCount; i++) {
+      final Future<Runnable> runnableFuture;
+      try {
+        runnableFuture = completionService.take();
+      } catch (InterruptedException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      }
+
+      final Runnable runnable;
+      try {
+        runnable = runnableFuture.get();
+      } catch (CancellationException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      } catch (ExecutionException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      } catch (InterruptedException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      }
+
+      runnable.run();
+    }
   }
 
   /**
-   * Start method for this poller, invoked by the CatalogFramework's blueprint when the framework is
-   * registered/installed. No logic is executed except for logging the framework name.
+   * May not return {@code null}
    *
-   * @param framework the catalog framework being started
-   * @param properties unused, but required by blueprint
+   * @throws NullPointerException if {@code source} is {@code null}
    */
-  public void start(CatalogFramework framework, Map properties) {
-    LOGGER.debug("Framework started for [{}]", framework);
-  }
+  abstract T getValue(final Source source) throws RuntimeException;
```

Comment:
❓Since these need to be overridden by the sub-classes, should we make them `protected`?

Reply:
✏️ We should clearly describe how the sub-classes are expected to implement those methods.